pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,351 @@
+// else return the unset bits.
+func (b *BitMap) Get(start uint32, end uint32, setBit bool) ([]*BitsRange, error) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	return b.getWithoutLock(start, end, setBit)
+}
+
+// Set sets or cleans the bits in range [start, end]. if setBit is true, set bits. else clean bits.
+func (b *BitMap) Set(start uint32, end uint32, setBit bool) error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	return b.setWithoutLock(start, end, setBit)
+}
+
+func (b *BitMap) getWithoutLock(start uint32, end uint32, setBit bool) ([]*BitsRange, error) {
+	if start > end {
+		return nil, fmt.Errorf("start %d should not bigger than %d", start, end)
+	}
+
+	if end > b.maxBitIndex {
+		return nil, fmt.Errorf("end %d should not bigger than %d", end, b.maxBitIndex)
+	}
+
+	rs := []*BitsRange{}
+	var lastRange *BitsRange
+
+	second64MinIndex := ((start >> 6) + 1) << 6
+	first64MaxIndex := end
+	if first64MaxIndex >= second64MinIndex {
+		first64MaxIndex = second64MinIndex - 1
+	}
+
+	last64MinIndex := (end >> 6) << 6
+
+	// get range of first uint64
+	arr := b.getRangeOfUint64(b.bm[start>>6], (start>>6)<<6, setBit)
+	appendArr, first, last := b.getAppendAndFirstLastRange(arr, start, first64MaxIndex)
+	if first != nil && first != last {
+		rs = append(rs, first)
+	}
+	rs = append(rs, appendArr...)
+	lastRange = last
+
+	for i := second64MinIndex; i < last64MinIndex; i += 64 {
+		arr := b.getRangeOfUint64(b.bm[i>>6], (i>>6)<<6, setBit)
+		if len(arr) == 0 {
+			continue
+		}
+		appendArr, first, last := b.getAppendAndFirstLastRange(arr, i, i+63)
+		appendOne, newLast := b.mergeOrAppend(lastRange, first)
+		if appendOne != nil {
+			rs = append(rs, appendOne)
+		}
+
+		lastRange = last
+		if first == last && newLast != nil {
+			lastRange = newLast
+		} else if newLast != nil {
+			rs = append(rs, newLast)
+		}
+
+		rs = append(rs, appendArr...)
+	}
+
+	// get range of last uint64
+	if last64MinIndex >= second64MinIndex {
+		arr = b.getRangeOfUint64(b.bm[end>>6], (end>>6)<<6, setBit)
+		if len(arr) > 0 {
+			appendArr, first, last := b.getAppendAndFirstLastRange(arr, last64MinIndex, end)
+			appendOne, newLast := b.mergeOrAppend(lastRange, first)
+			if appendOne != nil {
+				rs = append(rs, appendOne)
+			}
+
+			if newLast != nil {
+				rs = append(rs, newLast)
+			}
+
+			rs = append(rs, appendArr...)
+			if first != last && last != nil {
+				rs = append(rs, last)
+			}
+
+			lastRange = nil
+		}
+	}
+
+	if lastRange != nil {
+		rs = append(rs, lastRange)
+	}
+
+	return rs, nil
+}
+
+func (b *BitMap) getRangeOfUint64(x uint64, base uint32, setBit bool) []*BitsRange {
+	var (
+		start, end uint
+	)
+
+	if !setBit {
+		x = uint64(^x)
+	}
+
+	res := []*BitsRange{}
+
+	for {
+		if x == 0 {
+			break
+		}
+
+		// start is the value of the trailing zeros of x
+		start = uint(Ctz64(x))
+		// remove the trailing zeros of x and counts again
+		end = uint(Ctz64(uint64(^(x >> start)))) + start
+
+		res = append(res, &BitsRange{
```

Comment:
I think the `append` is a very time-consuming operation compared to the other ops in this function. It's better to remember the `previous end position` of the whole `BitMap`, compare to `current start position` and only execute `append` when the `previous end position` is different from `current start position`, otherwise assign the value of `current start position` to `previous end` then execute the next loop.

Reply:
Good suggestion! Done by it.